### ISSUE TITLE
Please remove the recommendation to rely on C==$14 as the main way of SGB detection

### ISSUE THREAD

**Issue by nitro2k01** [gbdev/pandocs#442]
Ok, time to be "that guy".

While it can be a neat trick to know that the SGB boot ROM leaves `C==$14`, it can lead to both false positives and false negatives in plausible real world scenarios. Therefore I would suggest restoring the recommendation to use `MLT_REQ` as the recommended way of detecting SGB functionality. Below are the ways in which it could lead to false detection, as well as addressing some of the things mentioned in that text.

## False positives

The impact of a false positive is that any attempt to transfer data using `_TRN` commands will be seen as garbage on the screen.

- One of the Gameboy clone boot ROMs will leave `C==$14`. By my naming conventions it is called `fortune_boot.bin` but it may exist in variously named clones consoles.
- The game may run on a real SGB, through the boot ROM, without access to SGB commands. In particular one common way that this could happen is using the EZFlash Jr cartridge. This cartridge has trouble running on a SGB because of reset line issues. A common workaround to this issue is to hit the reset button at a point which allows the boot ROM to run and the menu to show. However, since the stock firmware of the EZFJ runs first and doesn't have SGB support enabled in its header, communication with the SNES BIOS is locked out after that point, including in the case that the SGB CPU is reset.

## False negatives

The impact of a false negative is that SGB functionality such as borders and screen colorization are disabled, when they don't need to be.

- Some cartridge menus on older flash cartridges don't reset the game, but instead store a small trampoline in WRAM which they use to jump to `$100` after setting up the cartridge's multi ROM functionality to show the ROM in question. One such example is the EMS 64M USB cartridge if using multi ROM functionality. Such a menu will almost certainly not leave `C==$14` or care to leave different register values for SGB compared to DMG.
- Older emulators typically don't produce correct initial register values for SGB, instead typically giving values similar to those found on a DMG. On such emulators, SGB functionality will be disabled, even the game would otherwise run fine in the emulator in question.
- There's one scenario where even newer more accurate emulators can fail to produce `C==$14`, namely in a "SGB+GBC" mode if such a mode exists. Both BGB and mGBA have such a mode and both (not surprisingly) start with the exact initial registers of a GBC. In this mode, SGB functionality is locked out, and the game is shown with an empty, useless border around it instead.

The conclusion is that relying solely on `C==$14` for SGB detection is likely to cause real world irritation for a small but significant group of users of the software you create and should not be recommended as the main SGB detection mechanism.

## Claims made in the current text

> but this method is more complicated 

True, but not by much. If you consider that the purpose of detecting SGB mode is typically to send over a border and/or colorize the screen, which in itself takes some setting up, the added effort in programming the detection algorithm is minor. A better solution to this point would be to create a library/example code that a developer can use as is.

> and slower than checking the value of the A and C registers after startup. 

So, it is true that SGB commands take some time to execute, due to a recommended delay after each command. Nintendo recommends a 70000 M cycle, or ~4 frame delay. Do that twice during detection (once to activate multiple controller mode, and then to deactivate it) and you're looking at an 8 frame delay, or ~133 ms. However, consider that on real hardware, it was directly preceded by a forced delay of anything from a 3.11 s (GBC) to 5.59 s (DMG). Also combine this with the fact that most games are expected to show a copyright, title or similar static screen as the first thing, which can be loaded, before this delay and thus essentially absorbs it . Given this context, a one-time 133 ms delay before user input is accepted should barely be perceptible to a human, much less an annoyance.

**Comment by ISSOtm** [comment]
MLT_REQ detection also has some false negatives, such as if a previously-ran firmware ran a MLT_REQ detection itself, but did not reset it to singleplayer. If the environment cannot be trusted, there's not much that can be done.

Additionally, MLT_REQ detection is known to be *very* finicky: timings must be obeyed, registers must be polled in the right order, interrupts must be disabled... there's a lot of moving parts, and a lot that can go wrong. (*Pokémon Yellow* being a famous example of SGB detection gone right for the wrong reasons.)
This complexity is what the current article lumps under "but this method is more complicated".

Also, what are the forced delays you noted calculated from? I don't know of any reasons why such would be needed on DMG or CGB, much less why they are shorter on the latter.

Note also that games may not show a static screen at that point. For example, *Aevilia* plays an animation (clouds scrolling by) during boot-up, *Libbet* fades the Nintendo logo out, etc.

---

Ultimately it's a question of how much effort you're willing to put into supporting broken setups: using boot-up `c` is much simpler and still correct for a large majority of users, while `MLT_REQ` is much more error-prone, slower, and still not bulletproof.

SGB+CGB specifically is ill-defined, as it does not emulate any actual hardware configuration (there is not enough bandwidth to transfer 15bpp tiles, and not enough for a hypothetical 8bpp "hack" instead), so every implementation is bound to diverge somehow, at least unless emulators agree on and publish a spec.

**Comment by CasualPokePlayer** [comment]
> Also, what are the forced delays you noted calculated from? I don't know of any reasons why such would be needed on DMG or CGB, much less why they are shorter on the latter.

Forced "delay" is from the bootroms.

**Comment by nitro2k01** [comment]
> SGB+CGB specifically is ill-defined, as it does not emulate any actual hardware configuration (there is not enough bandwidth to transfer 15bpp tiles, and not enough for a hypothetical 8bpp "hack" instead), so every implementation is bound to diverge somehow, at least unless emulators agree on and publish a spec.

And to be clear on this point, it does not invent a new mode with 15 bpp or 8 bpp transfers. It simply behaves as if the hardware was a good old 2 bpp SGB, when SGB transfers are being requested. So if the game happens to not lock out the SGB code paths when GBC is detected, it will (with some probability) "just work".

**Comment by nitro2k01** [comment]
> Note also that games may not show a static screen at that point. For example, Aevilia plays an animation (clouds scrolling by) during boot-up, Libbet fades the Nintendo logo out, etc.

Oh, and another point about this. If you're planning to use a SGB border (which you likely are if you're detecting SGB) you need to display a static screen for a short while *anyway* for the data transfer of the border graphics. So this is something you need to plan your creative decisions around at any rate.

Aevilia is a GBC only game, so not really applicable. Libbet appears to simply detect SGB the traditional way (meaning the Nintendo logo stays static during detection, even on a DMG) and accept the extra delay that comes with that.

**Comment by pinobatch** [comment]
> > Libbet fades the logo out
> 
> Libbet appears to simply detect SGB the traditional way (meaning the Nintendo logo stays static during detection, even on a DMG)

Correct. *Libbet* detects SGB using `MLT_REQ` [while the logo is still up](https://github.com/pinobatch/libbet/blob/f9c71574c167b75cbedcae60a376713697191096/src/init.z80#L86) in order to decide whether to knock away the ®.

- GBC (also A+B+Select+Start): Clear tilemap first, no ® deflection
- DMG/SGB: Wait 30 frames, detect via `MLT_REQ`, and schedule ® deflection only if DMG

Perhaps ISSOtm was thinking of 144p Test Suite, which [detects SGB while fading out the logo](https://github.com/pinobatch/240p-test-mini/blob/740fded078e788d9e72923ec24ef2c39dbf91bb4/gameboy/src/init.z80#L112-L125).

1. Dark gray: Wait 8 frames while (presumably) various things in the SGB front end continue to initialize
2. Light gray: Detect SGB using `MLT_REQ`

**Comment by nitro2k01** [comment]
A new "broken" setup that would give false positives using this detection method just dropped. There are now "gb to fc" adapters being sold that allow you to play GB cartridges on a Famicom, much like a SGB. A [teardown](https://www.youtube.com/watch?v=cqMgXVjjUMI&t=3s) reveals that it's actually using a salvaged SGB chip, but presumably without being able to send SGB commands. (I have one of those units waiting in the mail for analysis.)

I still hold that the recommendation should be reverted to suggest using `MLT_REQ` as the main method of detecting SGB.

**Comment by avivace** [comment]
Pan Docs’ primary objective is to document the main Game Boy hardware - we don’t necessarily need to exhaustively cover every edge case involving unofficial hardware, clone boot ROMs, or flashcarts.

That said, I think it would still be beneficial to add a clear warning regarding the C == $14 detection method - specifically, to resize expectations around how “safe” it is, and to mention the main situations where it can lead to false positives or false negatives.

I would also suggest improving the section about the MLT_REQ-based detection method. Right now it’s framed primarily as “more complicated and slower,” which undersells it. We could improve its description, maybe provide an example ?

Ultimately, I'm in favour of a reword/change of tones/some additional information so that readers can make a (more) informed decision and weight the cost of the alternative, rather than simply assuming C==$14 as rock solid (as I believe is currently the case).

**Comment by nitro2k01** [comment]
As I see it, `MLT_REQ` is the "tried and tested" method of detection. If you want to recommend a different way of detection, it's on you to justify why it's better, and I don't think the reasons presented outweigh the potential issues. Everything I mentioned doesn't need to be documented in Pan Docs, rather those points are arguments for why `MLT_REQ` should be the standard recommended detection method.

`MLT_REQ` is framed as being slow and complicated, however the main reason for wanting to detect SGB in the first place is likely to send additional SGB commands which would be equally slow and complicated, if not more so. The main value for `C == $14` might be if you want to detect SGB and then not do anything with that information for some reason. For example if you're making a "write once and forget" library to detect the system and put it in a struct in memory.
